File: Computer_Science/Common_Algorithms/Sorting/Permutation_Sort/Implementation/Python/Permutation_sort.py
import random


def is_sorted(random_array):
    for i in range(1, len(random_array)):
        if random_array[i] < random_array[i - 1]:
            return False
    return True


def bogo_sort(random_array):
    while not is_sorted(random_array):
        random.shuffle(random_array)
    return random_array


def generate_random_array(size, min_val, max_val):
    return [random.randint(min_val, max_val) for _ in range(size)]


def main():
    size = 10
    min_val = 1
    max_val = 100
    random_array = generate_random_array(size, min_val, max_val)
    print(f"Unsorted Array: {random_array}")
    sorted_array = bogo_sort(random_array)
    print(f"sorted Array: {sorted_array}")


main()

File: Computer_Science/Common_Algorithms/Sorting/Permutation_Sort/Implementation/C/Permutation_Sort.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int is_sorted(int *a, int size) {
  for (int i = 0; i < size - 1; i++) {
    if (a[i] > a[i + 1]) {
      return 0;
    }
  }
  return 1;
}

void print_shuffled_array(int *arr, size_t size) {
  printf("[");
  for (int i = 0; i < size; i++) {
    printf("%d ", arr[i]);
  }
  printf("]\n");
}

void shuffle(int *arr, size_t size) {
  int temp, random;
  for (int i = 0; i < size; i++) {
    random = (int)((double)rand() / ((double)RAND_MAX + 1) * size);
    temp = arr[random];
    arr[random] = arr[i];
    arr[i] = temp;
  }
  print_shuffled_array(arr, size);
}

void bogo_sort(int *arr, size_t size) {
  while (!is_sorted(arr, size)) {
    shuffle(arr, size);
  }
}

int main() {
  int arr[10] = {60, 59, 14, 58, 61, 84, 78, 30, 9, 20};
  size_t size = sizeof(arr) / sizeof(arr[0]);
  srand(time(NULL));

  bogo_sort(arr, size);

  for (int i = 0; i < size; i++) {
    printf("%d ", arr[i]);
  }
}
